### PR TITLE
Adding py-snippets.

### PR DESCRIPTION
**Pull request by Xaldew**
### Brief summary of what the package does

This package consists of a small collection of snippets for use together with [yasnippet](https://github.com/joaotavora/yasnippet). As noted in AndreaCrotti/yasnippet-snippets/issues/383. The snippets in this particular package are generally more complex (or radical, as I call them) to be included in a general purpose collections (such as the official collection [yasnippet](https://github.com/AndreaCrotti/yasnippet-snippets)), so these were extracted to form a new package.

### Direct link to the package repository

https://github.com/Xaldew/yasnippet-radical-snippets

### Your association with the package

I'm the maintainer of this package.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback:
     - There is one warning in regards to `use-after-load`, but I would argue it is desired since this package is intended to be used with yasnippet.
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] ~I have confirmed some of these without doing them~

### CODE REVIEW

**Comment by riscy** [comment]
Thanks for this.  This appears to be 5 snippets across 2 languages, and some of them don't look that radical to me (yasnippet itself already has numpy docstrings builtin, so google docstrings don't seem like a stretch).  They look useful as personal configuration, but I'd be disinclined to add a small grab-bag of yasnippets like this to MELPA.  Instead I'd encourage trying to get these added to yasnippet itself (or other language-specific snippet packages).

As for the code itself -- regarding the package-lint issue, it crops up because simply loading a package should cause few side effects, and it's something pedantically push back against.  A couple more notes while I was reviewing this:

- Prefer the default recipe or use `:defaults`, if possible.
- Add a LICENSE file that GitHub can detect (e.g. no markup) if possible
  See: https://github.com/licensee/licensee
- python-mode/.yas_setup.el appears like it will declare several functions that exist outside the package's namespace (e.g. `python-split-args`, `python-args-to-google-docstring`).  Even though this is just "setup", these should use your package's prefix (`yasnippet-radical-snippets-`) as well.  It seems yasnippet itself doesn't seem to follow this, but yasnippet probably predates MELPA's stricter enforcement of prefixes
- Related, `python-args-to-google-docstring` looks like it came from [here](https://github.com/tjaartvdwalt/emacs-config/blob/322d7009e425c87c75bcce22e1771b390ddf96ae/functions/yas-setup.el) (but I may be wrong) -- so it would be ideal to give that author credit

**Comment by Xaldew** [comment]
> Thanks for this. This appears to be 5 snippets across 2 languages, and some of them don't look that radical to me (yasnippet itself already has numpy docstrings builtin, so google docstrings don't seem like a stretch). They look useful as personal configuration, but I'd be disinclined to add a small grab-bag of yasnippets like this to MELPA. Instead I'd encourage trying to get these added to yasnippet itself (or other language-specific snippet packages).

That's what I was thinking initially and I did actually suggest that in issue AndreaCrotti/yasnippet-snippets/issues/383, but they were deemed a bit too complicated for a general collection, something I personally agree with which is also the main reason I created a separate package for these.

> Regarding the package-lint issue, it crops up because simply loading a package should cause few side effects, and it's something I pedantically push back against.

That makes sense. I adapted the recommended setup/use-package snippet to include that piece of code instead to avoid these side-effects.

> Prefer the default recipe or use :defaults, if possible.

Should be fixed in the amended commit.


> Add a LICENSE file that GitHub can detect (e.g. no markup) if possible See: https://github.com/licensee/licensee

Fixed.


> python-mode/.yas_setup.el appears like it will declare several functions that exist outside the package's namespace (e.g. python-split-args, python-args-to-google-docstring). Even though this is just "setup", these should use your package's prefix (yasnippet-radical-snippets-) as well. It seems yasnippet itself doesn't seem to follow this, but yasnippet probably predates MELPA's stricter enforcement of prefixes

This ended up a bit wordy, but also fixed.

> Related, python-args-to-google-docstring looks like it came from here (but I may be wrong) -- so it would be ideal to give that author credit

Interesting, now that I think of it, I think that I did adapt some code from somewhere but I'm afraid I cannot remember exactly where... It is definitely not that one as that one though, as it was made available a few years after my first version of this [snippet](https://github.com/Xaldew/dotfiles/commit/e30b327b84172300285ba8b27e1e47deccd3328c#diff-c9d5b1ec64eb0859df8beb028b86d36db9b12961f084b5d109452c9521308fd6).

**Comment by riscy** [comment]
> That's what I was thinking initially ... https://github.com/AndreaCrotti/yasnippet-snippets/issues/383

Apologies, I didn't click through to that issue.  I think you didn't bring up the single cc-mode snippet in that issue, and since the snippet only has 5 fields maybe it would be pedestrian enough to include in the main yasnippet library?

Incidentally, it looks like there _is_ room for someone to come along and create a `py-snippets` package as well (this is your big chance :) which I think would be a much better way to package your 4 Python snippets, and could give other Python-snippet authors a place to consolidate and contribute.

My main pushback here is that I think it would be better for discoverability if snippets continued to tend to be organized by language.

**Comment by riscy** [comment]
Also the other issues look good now, so thanks for that.  I realize the namespacing makes things wordy, but the benefits are _really_ worth it.

**Comment by Xaldew** [comment]
>  Incidentally, it looks like there is room for someone to come along and create a py-snippets package as well (this is your big chance :) which I think would be a much better way to package your 4 Python snippets, and could give other Python-snippet authors a place to consolidate and contribute.

> My main pushback here is that I think it would be better for discoverability if snippets continued to tend to be organized by language.


Hmm, I do actually like the thought of creating such a `py-snippets` package. If you don't mind, lets keep this open a bit longer as I'm a bit strapped for time at the moment, but I'll see if the other snippets are acceptable in the general snippet collection before creating a new package

**Comment by riscy** [comment]
That sounds good to me.  I'll mark this awaiting-upstream in the meantime.

**Comment by riscy** [comment]
Friendly ping. :)

**Comment by Xaldew** [comment]
Hello!

Sorry about the (very) long delay, this just fell out of my priority list and was practically forgotten. That said, I've now authored a new repository (https://github.com/Xaldew/py-snippets) as we discussed which contains the Python snippets. Additionally, the recipe and this pull-request has been updated as needed. That said, feel free to redo any checks as necessary, it has been a while after all!

**Comment by riscy** [comment]
This looks good.  Thanks!
